### PR TITLE
[Vue][Web][SIMS #633] Quick navigation by Vue route from PIR and Student application forms

### PR DESCRIPTION
**Pull request by dheepak-aot**
Vue routing created for quick navigation from following 

- Program Information Request(Button created to navigate to "create a program" from this form)
- Student Application(Button created to navigate to "edit student profile" from this form)

Form IO wrapper component enhanced to accept and transmit custom events emitted from form.io

sfaa2022-23 updated and created sfaa2021-23 and sfaa2023-24 by copying sfaa2022-23 and changing the year respectively.

### CODE REVIEW

File: sources/packages/forms/src/form-definitions/sfaa2022-23.json [hunk 1]
```diff
@@ -14346,60 +14487,38 @@
                 "step": "any",
                 "integer": ""
               },
-              "key": "totalUnsubsidizedDayCareCostsDuringClassHoursForTheProposedStudyPeriodForDisabledInfirmedDependents12YearsOfAgeAndOlder",
+              "unique": false,
+              "errorLabel": "",
+              "key": "daycareCosts12YearsOrOver",
```
Comment (abschwenker) [inline sources/packages/forms/src/form-definitions/sfaa2022-23.json:14492]: This was a change I did so it is part of your form definition upload. I will not upload a separate definition

File: sources/packages/forms/src/form-definitions/sfaa2022-23.json [hunk 2]
```diff
@@ -14003,70 +14100,67 @@
...
-          "disabled": false,
-          "autofocus": false,
-          "dbIndex": false,
-          "customDefaultValue": "",
-          "calculateValue": "",
-          "calculateServer": false,
-          "widget": null,
-          "attributes": {},
-          "validateOn": "change",
+          "key": "haveDaycareCosts11YearsOrUnder",
```
Comment (abschwenker) [inline sources/packages/forms/src/form-definitions/sfaa2022-23.json:14126]: This was a change I did so it is part of your form definition upload. I will not upload a separate definition

File: sources/packages/forms/src/form-definitions/programinformationrequest.json [hunk 3]
```diff
@@ -2634,7 +2634,7 @@
               "value": ""
             }
           ],
-          "content": "Select the program that the applicant will be attending. If you know the program the student would like to attend does not exist in StudentAid BC you will need to create an approved program entry.",
+          "content": "Select the program that the applicant will be attending. If you know the program the student would like to attend does not exist in StudentAid BC you will need to create an approved program entry.\nClick the button below to create a program",
```
Comment (andrewsignori-aot) [inline sources/packages/forms/src/form-definitions/programinformationrequest.json:2637]: Please add a period in the end of the last sentence.

File: sources/packages/web/src/components/generic/formio.vue [hunk 4]
```diff
@@ -11,7 +11,7 @@ import ApiClient from "../../services/http/ApiClient";
 import FormUploadService from "../../services/FormUploadService";
 
 export default {
-  emits: ["submitted", "loaded", "changed"],
+  emits: ["submitted", "loaded", "changed", "custom"],
```
Comment (ann-aot) [inline sources/packages/web/src/components/generic/formio.vue:14]: can we change `custom` to something meaningful? like `customeCallback` or to something that makes more sense

File: sources/packages/web/src/components/generic/formio.vue [hunk 5]
```diff
@@ -97,6 +97,10 @@ export default {
       form.on("submit", (submision: any) => {
         context.emit("submitted", submision.data, form);
       });
+
+      form.on("customEvent", (event: any) => {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/generic/formio.vue:101]: Please add a type to the event defined as "any". As agreed in a previous PR, the idea is to create the types as needed for Form.IO. There is a simple type created under type/formio.ts.

File: sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue [hunk 6]
```diff
@@ -254,6 +255,14 @@ export default {
       }
     };
 
+    const customEventCallback = async (form: any, event: any) => {
+      if ("routeToStudentProfile" === event.type) {
```
Comment (ann-aot) [inline sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue:259]: maybe we can add enums for custom callback names in vue side

File: sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue [hunk 7]
```diff
@@ -254,6 +255,14 @@ export default {
       }
     };
 
+    const customEventCallback = async (form: any, event: any) => {
```
Comment (ann-aot) [inline sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue:258]: we know the type of `form`, we can try to add an interface or it

**Review by ann-aot — COMMENTED** [review]
Nice work with formio custom callback events 👍 . added few comments

**Review by andrewsignori-aot — COMMENTED** [review]
Overall looks good and it was a nice solution to emit and capture the custom event from Form.IO 👍 
Please take a look at the minor comments.

File: sources/packages/web/src/components/generic/formio.vue (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/generic/formio.vue:14]: All 4 in the list are callbacks actually, I would prefer "customEvent" over using the word callback here. 
What is your opinion?

Comment (andrewsignori-aot) [reply sources/packages/web/src/components/generic/formio.vue:14]: > All 4 in the list are events actually, I would prefer "customEvent" over using the word callback here.
> What is your opinion?

I agree with @dheepak-aot. Callback would be the methods that would be associated to these events. I would prefer to leave it as is.

File: sources/packages/forms/src/form-definitions/programinformationrequest.json (hunk 3 above)
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/programinformationrequest.json]: Period added.

File: sources/packages/web/src/components/generic/formio.vue (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/generic/formio.vue]: Changed from custom to customEvent

File: sources/packages/web/src/components/generic/formio.vue (hunk 5 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/generic/formio.vue]: Type added.

File: sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue (hunk 6 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue]: Enum created.

File: sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue]: Interface added

**Review by andrewsignori-aot — APPROVED** [review]
Looks good 👍

**Review by abschwenker — APPROVED** [review]
LGTM

**Review by ann-aot — APPROVED** [review]
LGTM 👍

**Review by guru-aot — APPROVED** [review]
LGTM

**Pull request merged**